21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md

```diff
@@ -117,14 +115,34 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
       -> StateT state monad a
       -> StateT state monad b
   map f (StateT g) = StateT (\state ->
-      let
-        (Tuple value state2) = g state
-        b = f value
-      in
-        pure (Tuple b state2)
+      (g state) >>= (\(Tuple value state2) ->
+        let b = f value in
+        pure $ Tuple b state2
+      )
     )
+
+-- Finally, we can using `do notation` in this case
+instance functor :: (Monad monad) => Functor (StateT state monad) where
+  map :: forall a b
+       . (a -> b)
+      -> StateT state monad a
+      -> StateT state monad b
+  map f (StateT g) = StateT \state ->
+    do
+      (Tuple value state2) <- g state
+      pure $ Tuple (f value) state2
 ```
 
+Actuclly, We don't even need the `monad` in `(StateT state monad)` to be a monad. What we did above is just take out the tuple from this monad and **map** it to a nother Tuple and than put the new tuple back to the same monad. Why don't we just **map** over it ? let's do it:
+```purescript
+instance functor :: (Functor f) => Functor (StateT state f) where
+  map :: forall a b
+      . (a -> b)
+      -> StateT state f a
+      -> StateT state f b
+  map aTob (StateT g) = StateT \state ->
+    map (\(Tuple a state2) -> Tuple (aTob a) state2) (g state)
+```
```

Comment:
While this is a good insight on your part, it's not my goal in this document to reveal that. I'd prefer either to not explain this insight here, or to explain it after the FAABM section.

Reply:
Yeah, I removed it.